File: kattis/contests/naq/naq21/k.py
m, n = map(int, input().split())
st = [*map(int, input().split())]

ps1 = [0]*m
ps1[0] = st[0]

ps2 = [0]*m
ps2[0] = st[-1]

for i in range(1, m):
    ps1[i] = ps1[i-1] + st[i]
    ps2[i] = ps2[i-1] + st[m-i-1]

ss = {0}
for p in ps1:
    ss.add(p)
for p in ps2:
    ss.add(p)

for i in range(m):
    for j in range(m-1-i):
        ss.add(ps1[i]+ps2[j])

for _ in range(n):
    q = int(input())
    print('Yes' if q in ss else 'No')

File: kattis/contests/naq/naq21/k.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    int m, n;
    cin >> m >> n;
    vector<int> st(m);
    for (int i = 0; i < m; i++) {
        cin >> st[i];
    }

    auto ps1 = vector<int>(m);
    ps1[0] = st[0];
    auto ps2 = vector<int>(m);
    ps2[0] = st[m - 1];

    for (int i = 1; i < m; i++) {
        ps1[i] = ps1[i - 1] + st[i];
        ps2[i] = ps2[i - 1] + st[m - i - 1];
    }

    unordered_set<int> ss = {0};
    for (int p: ps1) {
        ss.insert(p);
    }
    for (int p: ps2) {
        ss.insert(p);
    }
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < m - 1 - i; j++) {
            ss.insert(ps1[i] + ps2[j]);
        }
    }

    int q;
    while (cin >> q) {
        cout << (ss.count(q) ? "Yes" : "No") << endl;
    }
}